Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1014,6 +1014,35 @@ $ frida-trace -p 1372 -a "libjpeg.so!0x4793c"
 
 To learn more about all options for advanced usage, check the [documentation in the official Frida website](https://frida.re/docs/frida-trace/ "documentation").
 
+##### JNI Tracing
+
+As detailed in section [Reviewing Disassembled Native Code](#reviewing-disassembled-native-code), the first argument passed to every JNI function is a JNI interface pointer. This pointer contains a table of functions that allows native code to access the Android Runtime. Identifying calls to these functions can help with understanding library functionality, such as what strings are created or Java methods are called.
+
+[jnitrace](https://github.com/chame1eon/jnitrace "jnitrace") is a Frida based tool similar to frida-trace which specifically targets the usage of Android's JNI API by native libraries, providing a convenient way to obtain JNI method traces including arguments and return values.
+
+Like `frida-trace`, `jnitrace` can be installed via pip. However, `jnitrace` is only available for python3.
+
+```bash
+$ pip install jnitrace
+```
+
+You can easily install it by running `pip install jnitrace` and run it straightaway as follows:
+
+```bash
+$ jnitrace -l libnative-lib.so sg.vantagepoint.helloworldjni
+```
+
+<img src="Images/Chapters/0x05c/jni_tracing_helloworldjni.png" width="500px"/>
```

Comment:
Unfortunately, there is no accurate backtrace output for the helloworldjni app. I have thought of a couple of things I could do, let me know if you think any of them will work or if you have any suggestions:

- Crop the backtrace section from the screenshot
- Use the Frida fuzzy backtrace mode (shows output but the output is not correct)
- Show output from a different test application that does produce output

Reply:
Just disassembled the function in the HelloWorldJNI application and I believe it fails to find a backtrace because the "stringFromJNI" is not a "real" function. There is no ret instruction, it just movs the arguments into the correct registers and then jumps to the JNI method so that the JNI method can return directly back to the runtime.

I saw this in some of my test cases, now I think about it. I was able to force the compiler to stop this optimisation by having multiple JNI methods in the same function.

Anyway, for the backtrace my code does nothing special, it just calls the Frida API like as follows:

var bt = Thread.backtrace(this.context, Backtracer.ACCURATE)

That call returns a list of return addresses from the stack, for this app it returns an empty array for the reasons stated about. Using the Backtracer.FUZZY option I do get some output and it is *probably* detecting the correct return address into the runtime, but I haven't checked that to confirm.

If you look at the README on the jnitrace page you can see how the output looks when it does work.